# BAFU dataset extraction

Author: Thiago Nascimento (thiago.nascimento@eawag.ch)

This notebook is used to retrieve and concatenate the NADUF dataset. They present indeed different time-resolutions so not necessarly daily and hourly, but in different campaings. 

The output is one file per catchemnt (similar to the CAMELS_CH), with 44 columns:

* mean_discharge(m3/s)
* total_discharge(Mio m3)
* temperature_BAFU(°C)	
* pH_BAFU()	
* conductivity_25C_BAFU(µS/cm)	
* oxygen(mg/l)	
* oxygen_saturation(%)	
* pH_lab()	
* conductivity_20C_lab(µS/cm)
* total_hardness(mmol/l)
* alkalinity(mmol/l)	
* calcium(mg/l)	
* magnesium(mg/l)	
* nitrate(mg N/l)	
* total_nitrogen(mg N/l)	
* DRP(mg P/l)	
* total_phosphorus(mg P/l)	
* total_phosphorus_filtered(mg P/l)	
* chloride(mg/l)	
* fluoride(mg/l)	
* bromide(mg/l)	
* silicate(mg H4SiO4/l)	
* sulphate(mg SO4/l)	
* sodium(mg/l)	
* potassium(mg/l)	
* iron(mg/l)	
* TOC(mg C/l)	
* DOC(mg C/l)	
* suspended_material(mg/l)	
* chromium(µg/l)	
* zinc(µg/l)	
* copper(µg/l)	
* cadmium(µg/l)	
* lead(µg/l)	
* nickel(µg/l)	
* mercury(µg/l)	
* barium(µg/l)	
* strontium(µg/l)	
* arsenic(µg/l)	
* manganese(µg/l)	
* NP(µg/l)	
* NP1EO(µg/l)	
* NP2EO(µg/l)	
* NP3EO(µg/l)

## Requirements
**Python:**

* Python>=3.6
* Jupyter
* geopandas=0.10.2
* numpy
* os
* pandas=2.1.3
* scipy=1.9.0
* tqdm

Check the Github repository for an environment.yml (for conda environments) or requirements.txt (pip) file.

**Files:**

* 


**Directory:**

* Clone the GitHub directory locally
* Place any third-data variables in their respective directory.
* ONLY update the "PATH" variable in the section "Configurations", with their relative path to the EStreams directory. 


## References
* 
## Observations
* 

# Import modules

In [25]:
import pandas as pd
import geopandas as gpd
import numpy as np
import tqdm as tqdm
import os
import glob
import warnings
import re

# Configurations

In [2]:
# Only editable variables:
# Relative path to your local directory
PATH = ".."
# Suppress all warnings
warnings.filterwarnings("ignore")

* #### The users should NOT change anything in the code below here. 

In [3]:
# Non-editable variables:
PATH_OUTPUT = "results/interval_samples/"

# Set the directory:
os.chdir(PATH)

In [4]:
os.getcwd()

'c:\\Users\\nascimth\\Documents\\Thiago\\Eawag\\Python\\Scripts\\CAMELS_CH_chem'

# Import data
* FULL dataset

In [59]:
# Full dataset of interval (time-series)
dataset_naduf = pd.read_excel(r"data/NADUF/naduf_data_1981-2020_v5.xlsx")
dataset_naduf

naduf_id status_number                                          remark  \
0          1181             1                                             NaN   
1          1181             1                                             NaN   
2          1181             1                                             NaN   
3          1181             1                                             NaN   
4          1181             1                                             NaN   
...         ...           ...                                             ...   
14464      6169             1                                             NaN   
14465      6169             1                                             NaN   
14466      6169             1                                             NaN   
14467      6169             1                                             NaN   
14468      6169             9  "1. Woche: Ankunft im Labor erst am 16.12.20."   

       year            date_end    duration  mean_discharge  total_discharge  \
0      1982 1982-11-15 06:00:00  336.000000        5.313851         6.427634   
1      1982 1982-11-29 06:00:00  336.000000        9.046227        10.942316   
2      1982 1982-12-13 05:30:00  335.000000       10.864181        13.102202   
3      1982 1982-12-27 05:30:00  336.000000       27.653205        33.449317   
4      1983 1983-01-10 05:55:00  337.000000       12.789252        15.515920   
...     ...                 ...         ...             ...              ...   
14464  2020 2020-10-26 10:35:00  336.833333       10.570123        12.817331   
14465  2020 2020-11-09 10:15:00  335.666667       15.549123        18.789560   
14466  2020 2020-11-23 10:05:00  335.833333        6.439759         7.785669   
14467  2020 2020-12-07 10:05:00  336.000000        4.323554         5.229770   
14468  2020 2020-12-21 15:15:00  341.166667        8.327426        10.227744   

       temperature_BAFU  pH_BAFU  ...       zinc    copper   cadmium  \
0              7.271006      NaN  ...  10.406677  3.019255  0.073556   
1              4.826679      NaN  ...        NaN       NaN       NaN   
2              4.872490      NaN  ...        NaN       NaN       NaN   
3              3.792989      NaN  ...   26.11415  2.798573  0.025071   
4              3.064523      NaN  ...  28.858242  2.764272  0.026786   
...                 ...      ...  ...        ...       ...       ...   
14464          8.946462      NaN  ...        NaN       NaN       NaN   
14465          8.916996      NaN  ...        NaN       NaN       NaN   
14466          6.657077      NaN  ...        NaN       NaN       NaN   
14467          3.561146      NaN  ...        NaN       NaN       NaN   
14468          3.696536      NaN  ...        NaN       NaN       NaN   

           lead  nickel  mercury  barium  strontium arsenic manganese  
0      1.790373     NaN      NaN     NaN        NaN     NaN       NaN  
1           NaN     NaN      NaN     NaN        NaN     NaN       NaN  
2           NaN     NaN      NaN     NaN        NaN     NaN       NaN  
3      1.349287     NaN      NaN     NaN        NaN     NaN       NaN  
4      1.332136     NaN      NaN     NaN        NaN     NaN       NaN  
...         ...     ...      ...     ...        ...     ...       ...  
14464       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14465       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14466       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14467       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14468       NaN     NaN      NaN     NaN        NaN     NaN       NaN  

[14469 rows x 46 columns]

- Network

In [60]:
# Network NADUF
network_naduf = pd.read_excel(r"data/CAMELS_CH_chem_stations_short_v2.xlsx", sheet_name='naduf')
network_naduf

naduf_id             naduf_station naduf_water_body  longitude_LV03  \
0       1837             Porte du Scex            Rhône          557660   
1       1833                     Brugg             Aare          657000   
2       1835                 Mellingen            Reuss          662830   
3       1823               Andelfingen             Thur          693510   
4       1842                   Martina              Inn          830640   
5       1840                  Riazzino           Ticino          713670   
6       1832                   Hagneck             Aare          580680   
7       1827  Rheinfelden, Messstation            Rhein          627190   
8       1828     Münchenstein, Hofmatt             Birs          613570   
9       4409                 Appenzell           Sitter          749040   
10      1826                Laufenburg            Rhein          646000   
11      1830             Bern, Schönau             Aare          600710   
12      1825                  Rekingen            Rhein          667060   
13      1838         Chancy, Aux Ripes            Rhône          486600   
14      3717     Thörishaus, Sensematt            Sense          593350   
15      1836                Gebenstorf           Limmat          660530   
16      4879                Frauenfeld             Murg          709540   
17      1824              Rheinsfelden            Glatt          678040   
18      2064                   S-chanf              Inn          795800   
19      1831                  Gümmenen            Saane          585100   
20      1821    Diepoldsau, Rietbrücke            Rhein          766280   
21      1246                Neuenkirch  Sellenbodenbach          658530   
22      2078      Weil, Palmrainbrücke            Rhein          611740   
23      6169                     Emmen      Kleine Emme          663700   

    lattitude_LV03  area_camels_CH [km2]  hydro_naduf_distance [km]  \
0           133280                5239.4                        0.0   
1           259360               11681.3                        0.0   
2           252580                3385.8                        0.0   
3           272500                1701.6                        0.0   
4           197190                1937.5                        0.0   
5           113500                1613.3                       10.2   
6           211650                5111.9                        0.0   
7           267840               34479.4                        0.0   
8           263080                 887.3                        0.0   
9           244220                  74.4                        0.0   
10          267410               34010.3                        0.0   
11          198000                2941.2                        0.0   
12          269230               14719.1                        0.0   
13          112340               10312.7                        0.0   
14          193020                 351.2                        0.0   
15          260770                2412.0                        7.1   
16          269660                 213.3                        0.0   
17          269720                 417.4                        0.0   
18          165910                 616.0                        3.0   
19          199240                1881.3                        5.3   
20          250360                6297.4                        0.0   
21          218290                  10.4                        0.0   
22          272310               36404.4                        7.2   
23          213630                 478.3                        0.0   

   waterqual_id  
0            ok  
1            ok  
2            ok  
3            ok  
4            ok  
5            ok  
6            ok  
7            ok  
8            ok  
9            ok  
10           ok  
11           ok  
12           ok  
13           ok  
14           ok  
15           ok  
16           ok  
17           ok  
18        

In [61]:
len(dataset_naduf.naduf_id.unique())

29

In [62]:
dataset_naduf.naduf_id.unique()

array([1181, 1246, 1821, 1822, 1823, 1824, 1825, 1826, 1828, 1829, 1830,
       1831, 1832, 1833, 1835, 1836, 1837, 1838, 1840, 1842, 2044, 2045,
       2046, 2064, 2078, 3717, 4409, 4879, 6169], dtype=int64)

Observations
- 1827 is not present in the dataset. 

### Renaming the columns

In [63]:
column_rename_dict = {
    'naduf_id': 'naduf_id', 
    'status_number': 'status_number', 
    'remark':'remark' , 
    'year':'year', 
    'date_end':'date_end', 
    'duration':'duration',
    'mean_discharge': 'q_mean',
    'total_discharge': 'total_discharge(Miom3)',
    'temperature_BAFU': 'temp_water',
    'pH_BAFU': 'ph',
    'conductivity_25C_BAFU': 'EC',
    'oxygen': 'o',
    'oxygensaturation': 'OS',
    'pH_lab': 'phlab',
    'conductivity_20C_lab': 'EClab',
    'total_hardness': 'hardness',
    'alkalinity': 'alkalinity',
    'calcium': 'ca',
    'magnesium': 'mg',
    'nitrate': 'no3',
    'total_nitrogen': 'TN',
    'DRP': 'DRP',
    'total_phosphorus': 'TP',
    'total_phosphorus_filtered': 'TFP',
    'chloride': 'cl',
    'fluoride': 'f',
    'bromide': 'br',
    'silicate': 'h4sio4',
    'sulphate': 'so4',
    'sodium': 'na',
    'potassium': 'k',
    'iron': 'fe',
    'TOC': 'TOC',
    'DOC': 'DOC',
    'suspended_material': 'TSS',
    'chromium': 'cr',
    'zinc': 'zn',
    'copper': 'cu',
    'cadmium': 'cd',
    'lead': 'pb',
    'nickel': 'ni',
    'mercury': 'hg',
    'barium': 'ba',
    'strontium': 'sr',
    'arsenic': 'as',
    'manganese': 'mn'
}

In [64]:
# Rename columns based on the dictionary
dataset_naduf.rename(columns=column_rename_dict, inplace=True)

In [71]:
(dataset_naduf.duration/24).min()

8.854166666664241

In [53]:
# Convert to datetime:
dataset_naduf["date_end"] = pd.to_datetime(dataset_naduf["date_end"], format='%Y-%m-%d')

# Subtract hours (duration_column) from datetime_column
dataset_naduf['date_start'] = dataset_naduf['date_end'] - pd.to_timedelta(dataset_naduf['duration'], unit='h')

# Round the datetime to the nearest minute
dataset_naduf['date_start'] = dataset_naduf['date_start'].dt.round('S')

dataset_naduf

naduf_id status_number                                          remark  \
0          1181             1                                             NaN   
1          1181             1                                             NaN   
2          1181             1                                             NaN   
3          1181             1                                             NaN   
4          1181             1                                             NaN   
...         ...           ...                                             ...   
14464      6169             1                                             NaN   
14465      6169             1                                             NaN   
14466      6169             1                                             NaN   
14467      6169             1                                             NaN   
14468      6169             9  "1. Woche: Ankunft im Labor erst am 16.12.20."   

       year            date_end    duration     q_mean  \
0      1982 1982-11-15 06:00:00  336.000000   5.313851   
1      1982 1982-11-29 06:00:00  336.000000   9.046227   
2      1982 1982-12-13 05:30:00  335.000000  10.864181   
3      1982 1982-12-27 05:30:00  336.000000  27.653205   
4      1983 1983-01-10 05:55:00  337.000000  12.789252   
...     ...                 ...         ...        ...   
14464  2020 2020-10-26 10:35:00  336.833333  10.570123   
14465  2020 2020-11-09 10:15:00  335.666667  15.549123   
14466  2020 2020-11-23 10:05:00  335.833333   6.439759   
14467  2020 2020-12-07 10:05:00  336.000000   4.323554   
14468  2020 2020-12-21 15:15:00  341.166667   8.327426   

       total_discharge(Miom3)  temp_water  ph  ...        cu        cd  \
0                    6.427634    7.271006 NaN  ...  3.019255  0.073556   
1                   10.942316    4.826679 NaN  ...       NaN       NaN   
2                   13.102202    4.872490 NaN  ...       NaN       NaN   
3                   33.449317    3.792989 NaN  ...  2.798573  0.025071   
4                   15.515920    3.064523 NaN  ...  2.764272  0.026786   
...                       ...         ...  ..  ...       ...       ...   
14464               12.817331    8.946462 NaN  ...       NaN       NaN   
14465               18.789560    8.916996 NaN  ...       NaN       NaN   
14466                7.785669    6.657077 NaN  ...       NaN       NaN   
14467                5.229770    3.561146 NaN  ...       NaN       NaN   
14468               10.227744    3.696536 NaN  ...       NaN       NaN   

             pb   ni   hg  ba  sr  as  mn          date_start  
0      1.790373  NaN  NaN NaN NaN NaN NaN 1982-11-01 06:00:00  
1           NaN  NaN  NaN NaN NaN NaN NaN 1982-11-15 06:00:00  
2           NaN  NaN  NaN NaN NaN NaN NaN 1982-11-29 06:30:00  
3      1.349287  NaN  NaN NaN NaN NaN NaN 1982-12-13 05:30:00  
4      1.332136  NaN  NaN NaN NaN NaN NaN 1982-12-27 04:55:00  
...         ...  ...  ...  ..  ..  ..  ..                 ...  
14464       NaN  NaN  NaN NaN NaN NaN NaN 2020-10-12 09:45:00  
14465       NaN  NaN  NaN NaN NaN NaN NaN 2020-10-26 10:35:00  
14466       NaN  NaN  NaN NaN NaN NaN NaN 2020-11-09 10:15:00  
14467       NaN  NaN  NaN NaN NaN NaN NaN 2020-11-23 10:05:00  
14468       NaN  NaN  NaN NaN NaN NaN NaN 2020-12-07 10:05:00  

[14469 rows x 47 columns]

In [30]:
#dataset_naduf = dataset_naduf[['naduf_id', 'date',
#       'mean_discharge(m3/s)',
#       'temperature(°C)', 'pH(-)', 'conductivity_25C(µS/cm)',
#       'oxygen(mg/l)', 'oxygen_saturation(%)', 'pH_lab(-)',
#       'conductivity_20C_lab(µS/cm)', 'total_hardness(mmol/l)',
#       'alkalinity(mmol/l)', 'calcium(mg/l)', 'magnesium(mg/l)',
#       'nitrate(mgN/l)', 'total_nitrogen(mgN/l)', 'DRP(mgP/l)',
#       'total_phosphorus(mgP/l)', 'total_phosphorus_filtered(mgP/l)',
#       'chloride(mg/l)', 'fluoride(mg/l)', 'bromide(mg/l)',
#       'silicate(mgH4SiO4/l)', 'sulphate(mgSO4/l)', 'sodium(mg/l)',
#       'potassium(mg/l)', 'iron(mg/l)', 'TOC(mgC/l)', 'DOC(mgC/l)',
#       'suspended_material(mg/l)', 'chromium(µg/l)', 'zinc(µg/l)',
#       'copper(µg/l)', 'cadmium(µg/l)', 'lead(µg/l)', 'nickel(µg/l)',
#       'mercury(µg/l)', 'barium(µg/l)', 'strontium(µg/l)', 'arsenic(µg/l)',
#       'manganese(µg/l)']]
#dataset_naduf

In [54]:
dataset_naduf = dataset_naduf[[
    'naduf_id', 
    'date_start',
    'date_end', 
    'q_mean',
    'temp_water',
    'ph',
    'EC',
    'phlab',
    'EClab',
    'OS',
    'hardness',
    'alkalinity',
    'TN',
    'DRP',
    'TP',
    'TFP',
    'TOC',
    'DOC',
    'TSS',
    'as',
    'ba',
    'br',
    'ca',
    'cd',
    'cl',
    'cr',
    'cu',
    'fe',
    'f',
    'h4sio4',
    'hg',
    'k',
    'mg',
    'mn',
    'na',
    'ni',
    'no3',
    'o',
    'pb',
    'so4',
    'sr',
    'zn',
    ]]
dataset_naduf

naduf_id          date_start            date_end     q_mean  \
0          1181 1982-11-01 06:00:00 1982-11-15 06:00:00   5.313851   
1          1181 1982-11-15 06:00:00 1982-11-29 06:00:00   9.046227   
2          1181 1982-11-29 06:30:00 1982-12-13 05:30:00  10.864181   
3          1181 1982-12-13 05:30:00 1982-12-27 05:30:00  27.653205   
4          1181 1982-12-27 04:55:00 1983-01-10 05:55:00  12.789252   
...         ...                 ...                 ...        ...   
14464      6169 2020-10-12 09:45:00 2020-10-26 10:35:00  10.570123   
14465      6169 2020-10-26 10:35:00 2020-11-09 10:15:00  15.549123   
14466      6169 2020-11-09 10:15:00 2020-11-23 10:05:00   6.439759   
14467      6169 2020-11-23 10:05:00 2020-12-07 10:05:00   4.323554   
14468      6169 2020-12-07 10:05:00 2020-12-21 15:15:00   8.327426   

       temp_water  ph  EC     phlab       EClab  OS  ...        mg  mn  \
0        7.271006 NaN NaN  8.480000  367.066770 NaN  ...     6.561 NaN   
1        4.826679 NaN NaN  8.430766  356.188458 NaN  ...  5.283114 NaN   
2        4.872490 NaN NaN  8.356344  348.381529 NaN  ...  4.937076 NaN   
3        3.792989 NaN NaN  8.474643  338.992271 NaN  ...     4.617 NaN   
4        3.064523 NaN NaN  8.281437  274.674792 NaN  ...  4.842543 NaN   
...           ...  ..  ..       ...         ...  ..  ...       ...  ..   
14464    8.946462 NaN NaN  8.410000  317.222586 NaN  ...  5.476386 NaN   
14465    8.916996 NaN NaN  8.390000  313.269589 NaN  ...  5.290715 NaN   
14466    6.657077 NaN NaN  8.380000  342.636797 NaN  ...  6.274621 NaN   
14467    3.561146 NaN NaN  8.400031  383.225945 NaN  ...  6.829544 NaN   
14468    3.696536 NaN NaN  8.414994  368.479080 NaN  ...  6.458998 NaN   

             na   ni       no3   o        pb        so4  sr         zn  
0      5.164441  NaN  1.680668 NaN  1.790373         15 NaN  10.406677  
1      4.411181  NaN  1.521885 NaN       NaN  12.741208 NaN        NaN  
2      4.184406  NaN  1.353225 NaN       NaN  13.263437 NaN        NaN  
3       3.55214  NaN  1.687782 NaN  1.349287  10.492866 NaN   26.11415  
4      3.992816  NaN  1.466389 NaN  1.332136   11.32136 NaN  28.858242  
...         ...  ...       ...  ..       ...        ...  ..        ...  
14464  4.041707  NaN  0.983409 NaN       NaN  10.812015 NaN        NaN  
14465  3.676866  NaN  1.010804 NaN       NaN   9.252906 NaN        NaN  
14466  5.204035  NaN   1.22541 NaN       NaN  11.903467 NaN        NaN  
14467  8.752959  NaN  1.689263 NaN       NaN   12.92145 NaN        NaN  
14468  7.982762  NaN  1.599202 NaN       NaN  10.758361 NaN        NaN  

[14469 rows x 42 columns]

In [55]:
# Function to round numbers and preserve symbols
def round_values(val):
    if isinstance(val, str):  # Handle string values with symbols
        if val.startswith('>') or val.startswith('<'):
            symbol = val[0]  # Extract the symbol ('>' or '<')
            try:
                number = float(val[1:])  # Convert the rest to a float
                return f"{symbol}{round(number, 4)}"
            except ValueError:  # Handle cases where conversion might fail
                return val
        else:
            try:
                return str(round(float(val), 4))  # Round plain string numbers
            except ValueError:
                return val  # Return original value if conversion fails
    elif isinstance(val, (int, float)):  # Handle numeric values
        return round(val, 4)
    return val  # Return unchanged if it's neither string nor numeric

In [56]:
# Network CAMELS_CH_Chem
network_camels_ch_chem = pd.read_excel(r"data/CAMELS_CH_chem_stations_short_v2.xlsx", sheet_name='all_4')
#network_camels_ch_chem.set_index("basin_id", inplace=True)
network_camels_ch_chem

basin_id  bafu_id  naduf_id  nawa_id isot_id            hydro_station  \
0        2009   2009.0    1837.0   1837.0   NIO04            Porte du Scex   
1        2011   2011.0       NaN   4070.0     NaN                     Sion   
2        2016   2016.0    1833.0   1833.0   NIO02                    Brugg   
3        2018   2018.0    1835.0   1339.0     NaN                Mellingen   
4        2019   2019.0       NaN   1852.0   NIO01              Brienzwiler   
..        ...      ...       ...      ...     ...                      ...   
112      2617   2617.0       NaN      NaN     NaN                  Müstair   
113      2623   2623.0       NaN      NaN     NaN                 Oberwald   
114      2634   2634.0    6169.0   1181.0     NaN                    Emmen   
115      2635   2635.0       NaN      NaN     NaN        Einsiedeln, Gross   
116      2640      NaN       NaN   1504.0     NaN  Delémont, Pré-Guillaume   

    hydrowater_body     lon     lat  area_camels  ...  lon_naduf lat_naduf  \
0             Rhône  557660  133280       5239.4  ...   557660.0  133280.0   
1             Rhône  593770  118630       3372.4  ...        NaN       NaN   
2              Aare  657000  259360      11681.3  ...   657000.0  259360.0   
3             Reuss  662830  252580       3385.8  ...   662830.0  252580.0   
4              Aare  649930  177380        555.2  ...        NaN       NaN   
..              ...     ...     ...          ...  ...        ...       ...   
112             Rom  830800  168700        128.6  ...        NaN       NaN   
113           Rhone  669900  154075         93.3  ...        NaN       NaN   
114     Kleine Emme  663700  213630        478.3  ...   663700.0  213630.0   
115       Grossbach  700710  218125          8.9  ...        NaN       NaN   
116           Sorne  593380  245940        213.9  ...        NaN       NaN   

    area_naduf bafu_naduf_distance        nawa_station  lon_nawa  lat_nawa  \
0       5239.4                 0.0       Porte du Scex  557660.0  133280.0   
1          NaN                 NaN                Sion  593277.0  118449.0   
2      11681.3                 0.0               Brugg  657000.0  259360.0   
3       3385.8                 0.0          Gebenstorf  659450.0  258850.0   
4          NaN                 NaN  Brienzerseeeinlauf  646692.0  177000.0   
..         ...                 ...                 ...       ...       ...   
112        NaN                 NaN                 NaN       NaN       NaN   
113        NaN                 NaN                 NaN       NaN       NaN   
114      478.3                 0.0        Emmen-Littau  663917.0  213356.0   
115        NaN                 NaN                 NaN       NaN       NaN   
116        NaN                 NaN            Delémont  593800.0  246600.0   

        area_nawa bafu_nawa_distance remarks.1  
0     5239.402096                0.0       NaN  
1     3372.417040                0.0       NaN  
2    11681.282882                0.0       NaN  
3     3420.503458               10.0       NaN  
4      555.808970                3.3       NaN  
..            ...                ...       ...  
112           NaN                NaN       NaN  
113           NaN                NaN       NaN  
114    478.277165                0.6       NaN  
115           NaN                NaN       NaN  
116    213.930000                0.7       NaN  

[117 rows x 31 columns]

In [58]:
for code in tqdm.tqdm(network_naduf.naduf_id):
    
    dataset = dataset_naduf[dataset_naduf["naduf_id"] == code]
    dataset.set_index("date_start", inplace = True)
    dataset.drop(["naduf_id"], axis=1, inplace = True)
    
    dataset.index.name = "date_start"
    
    # Apply the function to the column
    dataset = dataset.applymap(round_values)

    # There are some non-numeric things in the columns, instead of NaNs
    #dataset = dataset.apply(pd.to_numeric, errors='coerce')
    
    # Here we take out the > or < before converting to a numeric value:
    #dataset = dataset.applymap(lambda x: str(x).replace('<', '') if isinstance(x, str) else x)
    #dataset = dataset.applymap(lambda x: str(x).replace('>', '') if isinstance(x, str) else x)

    # There are some non-numeric things in the columns, instead of NaNs
    #dataset = dataset.apply(pd.to_numeric, errors='coerce')

    #dataset = dataset.round(4)
    basin_id_name = str(network_camels_ch_chem[network_camels_ch_chem.naduf_id == code].loc[:, "basin_id"].values[0])

    dataset.to_csv(PATH_OUTPUT + "/NADUF/CAMELS_CH_Chem_naduf_"+str(basin_id_name)+".csv", encoding='latin')
    

100%|██████████| 24/24 [00:01<00:00, 22.52it/s]


In [17]:
dataset.columns

Index(['q_mean', 'temp_water', 'ph', 'EC', 'phlab', 'EClab', 'OS', 'hardness',
       'alkalinity', 'TN', 'DRP', 'TP', 'TFP', 'TOC', 'DOC', 'TSS', 'as', 'ba',
       'br', 'ca', 'cd', 'cl', 'cr', 'cu', 'fe', 'f', 'h4sio4', 'hg', 'k',
       'mg', 'mn', 'na', 'ni', 'no3', 'o', 'pb', 'so4', 'sr', 'zn'],
      dtype='object')

Observations
- We have 24 stations in total (one is empty: 1827)
- So far, the itnervals are variable (not resampled)

# End